tests/integration/security/sds_ingress/util/util.go

```diff
@@ -289,8 +290,61 @@ func SendRequestOrFail(ctx framework.TestContext, ing ingress.Instance, host str
 				func(resp client.ParsedResponses, err error) error {
 					// Check that the error message is expected.
 					if err != nil {
-						if !strings.Contains(err.Error(), exRsp.ErrorMessage) {
-							return fmt.Errorf("expected response error message %s but got %v",
+						// If expected error message is empty, but we got some error
+						// message then it should be treated as error when error message
+						// verification is not skipped. Error message verification is skipped
+						// when the error message is non-deterministic.
+						if !exRsp.SkipErrorMessageVerification && len(exRsp.ErrorMessage) == 0 {
+							return fmt.Errorf("unexpected error: %w", err)
+						}
+						if !exRsp.SkipErrorMessageVerification && !strings.Contains(err.Error(), exRsp.ErrorMessage) {
+							return fmt.Errorf("expected response error message %s but got %w",
+								exRsp.ErrorMessage, err)
+						}
+						return nil
+					}
+
+					return resp.CheckCode(strconv.Itoa(exRsp.ResponseCode))
+				})),
+	}
+
+	if callType == Mtls {
+		opts.Key = tlsCtx.PrivateKey
+		opts.Cert = tlsCtx.Cert
+	}
+
+	// Certs occasionally take quite a while to become active in Envoy, so retry for a long time (2min)
+	ing.CallWithRetryOrFail(ctx, opts, retry.Timeout(time.Minute*2))
+}
+
+func SendQUICRequestOrFail(ctx framework.TestContext, ing ingress.Instance, host string, path string,
```

Comment:
seems we can reduce duplication with SendRequestOrFail a lot

Reply:
I guess it was something similar originally. `SendRequestOrFail` is used in many places. However, I'll still change it even though the diff would be more - it would be trivial anyways.

Reply:
What I would do is:

```
SendRequestOrFail () { internalSend(protocol=TCP) }
SendQUICRequestOrFail() {internalSend(protocol=QUIC) }
internalSend() { .. existing logic ...}
```

Reply:
Ah! Right! This is neat.